### ISSUE TITLE
AVX512 compilation error on MSVC 19.16.27035.0 (VS 2017)

### ISSUE THREAD

**Issue by pitrou** [xtensor-stack/xsimd#421]
While trying to compile code using `xsimd::batch<uint32_t, 16>` with AVX512 enabled, I got these errors on AppVeyor:
https://ci.appveyor.com/project/pitrou/arrow/builds/37935012/job/5k8hyx5vm69f5bav#L1169

```
c:\projects\arrow\cpp\build\xsimd_ep\src\xsimd_ep-install\include\xsimd\types\xsimd_sse_conversion.hpp(90): error C3861: '_mm_cvtepu32_ps': identifier not found
c:\projects\arrow\cpp\build\xsimd_ep\src\xsimd_ep-install\include\xsimd\types\xsimd_avx_conversion.hpp(126): error C3861: '_mm256_cvtepu32_ps': identifier not found
c:\projects\arrow\cpp\build\xsimd_ep\src\xsimd_ep-install\include\xsimd\types\xsimd_avx512_double.hpp(452): error C2440: 'type cast': cannot convert from '__m512d' to '__m512i'
c:\projects\arrow\cpp\build\xsimd_ep\src\xsimd_ep-install\include\xsimd\types\xsimd_avx512_double.hpp(452): note: No constructor could take the source type, or constructor overload resolution was ambiguous
c:\projects\arrow\cpp\build\xsimd_ep\src\xsimd_ep-install\include\xsimd\types\xsimd_avx512_double.hpp(452): error C2660: '_mm512_maskz_and_epi64': function does not take 2 arguments
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\zmmintrin.h(1121): note: see declaration of '_mm512_maskz_and_epi64'
[etc.]
```

While generated, the code which is being compiled is really simple:
https://github.com/pitrou/arrow/blob/xsimd-bpacking/cpp/src/arrow/util/bpacking_simd512_generated.h

**Comment by pitrou** [comment]
Here is a similar issue with MSVC 19.28.29337.0 (VS 2019):
https://github.com/pitrou/arrow/runs/1972721711?check_suite_focus=true#step:7:777

**Comment by pitrou** [comment]
Some thought after looking at the first few errors and checking in the online Intel intrinsics guide:
* `_mm_cvtepu32_ps` and `_mm256_cvtepu32_ps` indeed don't exist
* `_mm512_maskz_and_epi64` takes 3 arguments not 2
* `_mm512_and_epi64` takes 2 arguments not 1
* ...

It seems that somehow xsimd is assuming to find weird intrinsics that are not officially documented by Intel...?

**Comment by serge-sans-paille** [comment]
@pitrou the github link to `bpacking_simd512_generated.h` is no longer valid, could you share the faulty source code again, or provide an equivalent reproducer?

**Comment by pitrou** [comment]
@serge-sans-paille Ah, right, the name of the git branch changed, sorry: https://github.com/pitrou/arrow/blob/ARROW-11839-xsimd-bpacking/cpp/src/arrow/util/bpacking_simd512_generated.h

**Comment by serge-sans-paille** [comment]
> Some thought after looking at the first few errors and checking in the online Intel intrinsics guide:
> 
>     * `_mm_cvtepu32_ps` and `_mm256_cvtepu32_ps` indeed don't exist

They actually do ( see https://officedaytime.com/simd512e/simdimg/convert.php?f=vcvtudq2ps)
Under Linux, they are defined in `avx512vlintrin.h`, see https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/avx512vlintrin.h#L1487 for the coresponding intrinsic.

Using a more recent version of MSVC should fix the issue, as showcased by https://godbolt.org/z/fGq6hKE18

Can you share the compiler flags you're using?

**Comment by JohanMabille** [comment]
`xsimd` is supposed to work starting with VS2015, so I think we should provide a workaround here.

**Comment by pitrou** [comment]
> They actually do ( see https://officedaytime.com/simd512e/simdimg/convert.php?f=vcvtudq2ps)

That's not an official Intel site.
If you look for them in the official Intrinsics guide, you won't find them: https://software.intel.com/sites/landingpage/IntrinsicsGuide/#

So for some reason, non-official docs and Linux headers have an intrinsic that's neither in the official Intel docs nor in the MSVC headers. I don't know why.

As for compiler flags, I tried `/arch:AVX512`. Perhaps something more specific must be used?

**Comment by pitrou** [comment]
`/arch:AVX512` is supposed to enable all desired extensions according to https://docs.microsoft.com/fr-fr/cpp/build/reference/arch-x64?view=msvc-160

**Comment by pitrou** [comment]
I took at the header (`zmmintrin.h`) provided with Visual Studio 2017:
* `_mm_cvtepu32_ps` is not defined
* `_mm_cvtepu32_pd` is defined as:
```c
#define _mm_cvtepu32_pd(v1) \
        _mm_maskz_cvtepu32_pd((__mmask8)_MM_K0_REG, (v1))
```
(but `_mm_maskz_cvtepu32_ps` doesn't exist anyway)
* `_mm512_cvtepu32_ps` is defined as:
```c
#define _mm512_cvtepu32_ps(v) \
        _mm512_cvt_roundepu32_ps((v), _MM_FROUND_CUR_DIRECTION)
```

**Comment by pitrou** [comment]
Also, looks like this is not the only case of missing intrinsics:
https://developercommunity.visualstudio.com/t/missing-intrinsic-definitions-in-zmmintrinh/828399

**Comment by serge-sans-paille** [comment]
Thanksfor the detailed information. Note that the latest version of MSVC, as hinted by the godbolt link above, seems to include support for these non-official intrinsics. @JohanMabille should we implement a portability layer with older MSVC version?

**Comment by pitrou** [comment]
More spelunking:
* `_mm512_ceil_ps(v)` should probably be a shortcut for `_mm512_roundscale_ps(v, _MM_FROUND_TO_POS_INF)`
* ditto for `_mm512_ceil_pd` to `_mm512_roundscale_pd`
* similarly, `_mm512_floor_ps(v)` -> `_mm512_roundscale_ps(v, _MM_FROUND_TO_NEG_INF)`
* ditto for `_mm512_floor_pd`

**Comment by pitrou** [comment]
As for `float <-> uint32` conversions, I think it would be ok to leave them undefined on MSVC. It's probably not a common use case.

**Comment by JohanMabille** [comment]
@pitrou Thanks for detailed information!
@serge-sans-paille For the sake of completeness I would say yes ;)

**Comment by pitrou** [comment]
Also, notice this issue's title and the various logs: the errors are on VS 2017 and VS 2019.

**Comment by serge-sans-paille** [comment]
@pitrou can you check if the following fails in your setup:

```
cmake  -D DOWNLOAD_GTEST=1 -D ENABLE_XTL_COMPLEX=1  -D CMAKE_BUILD_TYPE=Release  -D CMAKE_CXX_FLAGS=/arch:AVX512 .
cmake --build . --target xtest
```

**Comment by serge-sans-paille** [comment]
I'm trying to setup CI that reproduces your issue but I'm failing at it...

**Comment by pitrou** [comment]
Have you tried https://github.com/xtensor-stack/xsimd/pull/423 ?

**Comment by serge-sans-paille** [comment]
**now** I feel ashamed. Thanks!

**Comment by pitrou** [comment]
Fixed by #451
